crates/goose/src/agent.rs

```diff
@@ -127,77 +133,176 @@ impl Agent {
+            &system_prompt,
+            &messages,
+            &tools,
+            &resources,
+            Some("gpt-4"),
+        );
 
-    // Rewrite the exchange as needed after each tool call
-    async fn rewrite_messages_on_tool_response(
-        &self,
-        messages: &mut Vec<Message>,
-        pending: Vec<Message>,
-        status: String,
-    ) -> AgentResult<()> {
-        // Remove the last two messages (status and tool response)
-        messages.pop();
-        messages.pop();
-
-        // Append the pending messages
-        messages.extend(pending);
-
-        // Add back a fresh status and tool response
+        let mut status_content: Vec<String> = Vec::new();
+
+        if approx_count > target_limit {
+            println!("[WARNING] Token budget exceeded. Current count: {} \n Difference: {} tokens over buget. Removing context", approx_count, approx_count - target_limit);
+
+            // Get token counts for each resourcee
+            let mut system_token_counts = HashMap::new();
+
+            // Iterate through each system and its resources
+            for (system_name, resources) in &resource_content {
+                let mut resource_counts = HashMap::new();
+                for (uri, (_resource, content)) in resources {
+                    let token_count = token_counter.count_tokens(&content, Some("gpt-4")) as u32;
+                    resource_counts.insert(uri.clone(), token_count);
+                }
+                system_token_counts.insert(system_name.clone(), resource_counts);
+            }
+            // Sort resources by priority and timestamp and trim to fit context limit
+            let mut all_resources: Vec<(String, String, Resource, u32)> = Vec::new();
+            for (system_name, resources) in &resource_content {
+                for (uri, (resource, _)) in resources {
+                    if let Some(token_count) = system_token_counts.get(system_name).and_then(|counts| counts.get(uri)) {
+                        all_resources.push((
+                            system_name.clone(),
+                            uri.clone(),
+                            resource.clone(),
+                            *token_count
+                        ));
+                    }
+                }
+            }
+
+            // Sort by priority (high to low) and timestamp (newest to oldest)
+            all_resources.sort_by(|a, b| {
+                let priority_cmp = b.2.priority.cmp(&a.2.priority);
+                if priority_cmp == std::cmp::Ordering::Equal {
+                    b.2.timestamp.cmp(&a.2.timestamp)
+                } else {
+                    priority_cmp
+                }
+            });
+
+            // Remove resources until we're under target limit
+            let mut current_tokens = approx_count;
+
+            while current_tokens > target_limit && !all_resources.is_empty() {
+                if let Some((system_name, uri, _, token_count)) = all_resources.pop() {
```

Comment:
What if one system ends up taking disproportionate number of tokens so another system ends up not being able to load anything into state? Should Goose set limits on how much each system can consume in its state? Or should that just result in an expected error if some systems can't be loaded?

Reply:
yea, i think more system configuration is needed for situations like setting priorities etc.. see previous comment